### PR TITLE
LOTTIE - Refactored code, added support for methods and callbacks

### PR DESCRIPTION
**Pull request by hemish11**
Things done in this PR:
- Refactored some code
- Added support for forward(), reverse(), reset() and stop() methods to the Lottie widget.
- Added onForward, onReverse, onComplete, onStop callbacks to the Lottie widget.
- Added autoPlay parameter.

### CODE REVIEW

**Comment by kmahmood74** [comment]
@hemish11 always mention the url to the ticket or ticket number with the pound sign in all your PRs. Without it we won't know which ticket you are resolving. 

Also the first thing in your commit should be the URL to the ticket. Github auto-links it with the ticket when you do that. 

for this one, it's https://github.com/EnsembleUI/ensemble/issues/914

File: lib/widget/lottie/lottie.dart [hunk 1]
```diff
@@ -28,7 +26,22 @@ class EnsembleLottie extends StatefulWidget
...
   @override
   Map<String, Function> methods() {
-    return {};
+    return {
+      // Method to start animation in forward direction
+      'forward': () {
+        if (_controller.repeat) {
+          _controller.lottieController!.repeat();
+        } else {
+          _controller.lottieController!.repeat();
```
Comment (kmahmood74) [inline lib/widget/lottie/lottie.dart:35]: in both cases you are calling repeat, what's the point of else?

File: lib/widget/lottie/native/lottiestate.dart [hunk 2]
```diff
@@ -8,65 +9,139 @@ import 'package:ensemble/widget/widget_util.dart';
...
+      AnimationStatus.forward: controller.onForward,
+      AnimationStatus.reverse: controller.onReverse,
+      AnimationStatus.dismissed: controller.onStop,
+      AnimationStatus.completed: controller.onComplete,
+    };
+
+    controller.lottieController!.addStatusListener(
+      (status) {
+        if (animationStatusActionMap[status] != null) {
+          ScreenController().executeAction(
```
Comment (kmahmood74) [inline lib/widget/lottie/native/lottiestate.dart:31]: this code that executes the action should be inside LottieController. That's what the controller is for. Also that way you can re-use most of the code between web and native

File: lib/widget/lottie/web/lottiestate.dart [hunk 3]
```diff
@@ -14,115 +18,201 @@ import 'dart:js' as js;
...
+    final animationStatusActionMap = <AnimationStatus, EnsembleAction?>{
+      AnimationStatus.forward: controller.onForward,
+      AnimationStatus.reverse: controller.onReverse,
+      AnimationStatus.dismissed: controller.onStop,
+      AnimationStatus.completed: controller.onComplete,
+    };
+
+    controller.lottieController!.addStatusListener(
+      (status) {
+        if (animationStatusActionMap[status] != null) {
```
Comment (kmahmood74) [inline lib/widget/lottie/web/lottiestate.dart:41]: this should be in LottieController, see my previous comment in the native LottieState

File: lib/widget/lottie/web/lottiestate.dart [hunk 4]
```diff
@@ -14,115 +18,201 @@ import 'dart:js' as js;
...
-    double height = widget.controller.height?.toDouble() ?? 250;
-    String _id = widget.controller.id ?? id;
-    String repeat = (widget.controller.repeat == null)
-        ? 'true'
-        : widget.controller.repeat!.toString();
-    String tag = '<div id="$_id"></div>';
-    if (widget.controller.width == null || widget.controller.height == null) {
-      print(
-          'Lottie widget must have explicit width and height for html renderer in the browser (this includes the preview in the editor). You do not need to specify width or height for native apps or for canvaskit renderer. Defaulting to $width for width and $height for height');
+  Widget buildLottieHtml(BoxFit? fit, LottieController controller) {
```
Comment (kmahmood74) [inline lib/widget/lottie/web/lottiestate.dart:87]: widget.controller is available everywhere inside the state. Why then create a variable called controller and then pass it around. 

Most of these changes are unnecessary and make it very hard to actually review the change. Please focus only on the actual change. Remove these unnecessary changes and update your PR

File: lib/widget/lottie/web/lottiestate.dart [hunk 5]
```diff
@@ -14,115 +18,201 @@ import 'dart:js' as js;
...
-            String script =
-                'bodymovin.loadAnimation({container: document.getElementById("$_id"),renderer: "svg",loop: $repeat,autoplay: true,path: "$c"});';
-            if (widget.controller.onTap != null) {
-              script +=
-                  'document.getElementById("$_id").addEventListener("click",() => handleMessage("$_id",""));';
-            }
-            return script;
-          },
-          size: Size(width, height));
+        id: id,
```
Comment (kmahmood74) [inline lib/widget/lottie/web/lottiestate.dart:106]: none of these changes are needed, please cleanup and minimize to only the changes that are required

File: lib/widget/lottie/web/lottiestate.dart [hunk 6]
```diff
@@ -14,115 +18,201 @@ import 'dart:js' as js;
...
+      );
     }
-    return blankPlaceholder();
+
+    return blankPlaceholder(controller);
   }
 
-  Widget buildLottieCanvas(BoxFit? fit) {
-    String source = widget.controller.source.trim();
+  Widget buildLottieCanvas(BoxFit? fit, LottieController controller) {
```
Comment (kmahmood74) [inline lib/widget/lottie/web/lottiestate.dart:136]: same issue, no reason to pass controller when it is available as widget.controller

File: lib/widget/lottie/web/lottiestate.dart [hunk 7]
```diff
@@ -14,115 +18,201 @@ import 'dart:js' as js;
...
+        },
+      );
     }
-    return blankPlaceholder();
+
+    return blankPlaceholder(controller);
   }
 
-  Widget placeholderImage() {
+  Widget placeholderImage(LottieController controller) {
```
Comment (kmahmood74) [inline lib/widget/lottie/web/lottiestate.dart:186]: once again, same issue, cleanup please

**Review by kmahmood74 — CHANGES_REQUESTED** [review]

**Comment by hemish11** [comment]
@kmahmood74 Please review this now

**Review by kmahmood74 — APPROVED** [review]
looks good, make sure to test it as follows - 

1. Web - test it with canvasKit AND with html renderer. Compile it with htmlrenderer option and test as well
2. iOS native
3. Android native

**Comment by hemish11** [comment]
@kmahmood74 @snehmehta Finally ready for review!

File: lib/widget/lottie/lottie.dart [hunk 8]
```diff
@@ -28,7 +30,66 @@ class EnsembleLottie extends StatefulWidget
 
   @override
   Map<String, Function> methods() {
-    return {};
+    if (kIsWeb) {
+      // A little hacky way to check if html renderer is used as only html render would have the lottieController null.
+      final bool isHtml = _controller.lottieController != null;
```
Comment (kmahmood74) [inline lib/widget/lottie/lottie.dart:35]: yes this is a really hacky way. Also I see that you use _controller.lottieController! everywhere.

**Comment by kmahmood74** [comment]
@hemish11 I added some comments

File: lib/widget/lottie/lottie.dart [hunk 9]
```diff
@@ -28,7 +30,66 @@ class EnsembleLottie extends StatefulWidget
...
+          } else {
+            _controller.lottieAction!.stop();
+          }
+        },
+      };
+    }
+
+    return {
+      // Method to start animation in forward direction
+      'forward': () {
```
[the remark this answers is not in the document]
Comment (hemish11) [reply lib/widget/lottie/lottie.dart:79]: There is a difference, lottieController is a normal AnimationController and is passed to lottie widget directly and thus controls animation whereas lottieAction has its methods which are coded specifically for the HTML renderer. I had to do so as HTML render does all the things by running javascript code and thus had to segregate all the things.

File: lib/widget/lottie/lottie.dart (hunk 8 above)
Comment (kmahmood74) [reply lib/widget/lottie/lottie.dart:35]: @hemish11 this is just super hacky, please make it cleaner so anyone working on the code later can make sense of it

Comment (hemish11) [reply lib/widget/lottie/lottie.dart:35]: The simpler way that I first tried to do was to check by this condition: `js.context['flutterCanvasKit'] != null;` but for this, import `'dart:js'` as js; needs to be imported in the main lottie.dart file which is not possible as the app starts to crash when importing the web-only library when it runs on mobile. It has been imported in web/lottiestate.dart but as it only runs when it is in web there is no issue with that import. Also currently I'm adding more comments in the code so that its easy to understand.

**Comment by hemish11** [comment]
@kmahmood74 Added a few comments in code

**Review by kmahmood74 — APPROVED** [review]
